商品購入機能の実装

fujiyasu0522 opened the pull request:
# What
　商品購入機能の実装
# Why
　商品の購入をするため

テストコード
https://gyazo.com/a943b0f213a2142a43df42f8eedbe630

fujiyasu0522 committed:
商品購入機能の実装

テストコード
https://gyazo.com/a943b0f213a2142a43df42f8eedbe630

KenyaMatsumoto reviewed (commented):
レビューご提出ありがとうございます！
恐れ入りますが、こちらのプルリクエストで実際の画面や機能ができているか確認させていただきたいので、下記キャプチャ画像をご提出いただけますか？
その後、再度レビュー依頼フォームにてご提出をお願いいたします。

・購入に成功する動画
・購入が完了したら、トップページまたは購入完了ページに遷移する動画
・購入に失敗してエラーが表示させる動画
・ログアウト状態のユーザーが商品購入ページに遷移しようとすると、ログインページに遷移する動画
・ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移する動画
・出品者・出品者以外にかかわらず、ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画
・ログアウト状態のユーザーが、URLを直接入力して売却済み商品の商品情報編集ページへ遷移しようとすると、ログインページに遷移する動画

fujiyasu0522 committed:
出品者でも購入済みは編集できない

テストコードの実装
https://gyazo.com/a943b0f213a2142a43df42f8eedbe630

・購入に成功する動画
https://gyazo.com/b7d890e322a590c4765d1963cda87324

・購入が完了したら、トップページまたは購入完了ページに遷移する動画
https://gyazo.com/b7d890e322a590c4765d1963cda87324

・購入に失敗してエラーが表示させる動画
https://gyazo.com/df414cab7bf4e245e31184b030ecff05

・ログアウト状態のユーザーが商品購入ページに遷移しようとすると、ログインページに遷移する動画
https://gyazo.com/9118414bbc36d98edbb08c73da3ae610

・ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移する動画
https://gyazo.com/6009dcf0d864e499e989aa7cb16bab24

・出品者・出品者以外にかかわらず、ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画

出品者以外
https://gyazo.com/6967cefa409eac20307baefef673df46

出品者
https://gyazo.com/fa70d94e924b21f93f1eafd8b7b85283

・ログアウト状態のユーザーが、URLを直接入力して売却済み商品の商品情報編集ページへ遷移しようとすると、ログインページに遷移する動画
https://gyazo.com/5d68d1e7912fabfb493c5c45906fcaf7

tatsuki-kato commented on app/controllers/items_controller.rb:
```diff
@@ -24,6 +24,9 @@ def create
   end
 
   def edit
+    if @item.order.present?
+      redirect_to root_path
+    end
```
```rb
if @item.order.present?
  redirect_to root_path
end
```
こちらの記述をメソッド化してbefore_actionで呼び出せるようにし、update、destroyアクションにも適用しましょう。
そうすることでセキュリティを高めることができるためです。

tatsuki-kato commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,45 @@
+class OrdersController < ApplicationController
+  before_action :authenticate_user!
+  before_action :set_item, only: [:index, :create]
+  before_action :sold_out_item, only: [:index]
```
```rb
before_action :sold_out_item, only: [:index]
```
セキュリティを高めるため、こちらの記述にcreateアクションを追加しましょう。

tatsuki-kato commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,45 @@
+  before_action :sold_out_item, only: [:index]
+
+  def index
+    if current_user.id == @item.user_id
+      redirect_to root_path
+    else
```
```rb
if current_user.id == @item.user_id
  redirect_to root_path
~(略)~
```
出品者が購入ページに遷移しようとするとトップページにリダイレクトする上記の記述ですが、createアクションにも適用できるようにメソッド化し、before_actionで呼び出せるようにしましょう。

tatsuki-kato commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,45 @@
+  def create
+    @order_form = OrderForm.new(order_params)
+    if @order_form.valid?
+      Payjp.api_key = "sk_test_dedca0a8be8dd120cb297e8e"
```
```rb
 Payjp.api_key = "****************"
```
こちらですが、payjpのキーに関しては、環境変数に定義して、
そこから呼び出すような仕組みにしましょう。
セキュリティの観点から見ても、githubにpushされるファイルに直接記述するのは好ましくありません。

tatsuki-kato commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,45 @@
+    @order_form = OrderForm.new(order_params)
+    if @order_form.valid?
+      Payjp.api_key = "sk_test_dedca0a8be8dd120cb297e8e"
+      Payjp::Charge.create(
+        amount: @item.price,
+        card: order_params[:token],
+        currency: "jpy"
+      )
```
```rb
 Payjp.api_key = "****************"
Payjp::Charge.create(
        amount: @item.price, 
       card: order_params[:token],
        currency: "jpy"      )
```
こちらの記述ですが、セキュリティを高めるため、private以下にメソッド化して呼び出す形にしましょう。

tatsuki-kato commented on app/javascript/card.js:
```diff
@@ -0,0 +1,34 @@
+    Payjp.createToken(card, (status, response) => {
+      if (status == 200) {
+        const token = response.id;
+        console.log(token);
```
```js
console.log(token);
```
こちらの記述は削除しましょう。
理由：開発時の挙動確認のための記述なので、実装後は必要ないためです。

tatsuki-kato commented on app/models/order_form.rb:
```diff
@@ -0,0 +1,17 @@
+    validates :block_name
+    validates :postal_code,      format: { with: /\A\d{3}[-]\d{4}\z/ }
+    validates :delivery_area_id, numericality: { other_than: 0 }
+    validates :phone_number,     format: { with: /\A\d{11}\z/ }
+  end
```
user_idとitem_idにも必須であるバリデーション(presence: true)を記述しましょう。
今まではモデル間のアソシエーションを組むことで、自動的にpresence: trueが働いておりました。
 ただ、フォームオブジェクトでは直接アソシエーションを組んでいないためにバリデーションが必要となります。

tatsuki-kato commented on spec/models/order_form_spec.rb:
```diff
@@ -0,0 +1,72 @@
+          expect(@order_form.errors.full_messages).to include("Postal code is invalid")
+        end
+
+        it 'delivery_area_idが空だと購入できない' do
+          @order_form.delivery_area_id = nil
+          @order_form.valid?
+            expect(@order_form.errors.full_messages).to include("Delivery area can't be blank")
+        end
```
以下のテストを追加しましょう。
```
delivery_area_idが0だと購入できない
```

tatsuki-kato commented on spec/models/order_form_spec.rb:
```diff
@@ -0,0 +1,72 @@
+          @order_form.phone_number = " 123456789123"
+          @order_form.valid?
+          expect(@order_form.errors.full_messages).to include("Phone number is invalid")
+        end
+      end
```
モデルにuser_idとitem_idのバリデーションを追加した後は、テストにも以下の項目を実装しましょう！
```
・user_idが空では購入できないこと
・item_idが空では購入できないこと
```

tatsuki-kato commented on spec/models/order_form_spec.rb:
```diff
@@ -0,0 +1,72 @@
+    describe '購入情報の保存' do
+
+      context '購入情報がうまくいく時' do
+        it '全ての項目が入力されていれば購入ができる' do
+          expect(@order_form).to be_valid
+        end
```
正常系パターンのテスト項目を追記しましょう！
```
・建物名が抜けていても登録できること
```

tatsuki-kato commented on spec/models/order_form_spec.rb:
```diff
@@ -0,0 +1,72 @@
+          expect(@order_form.errors.full_messages).to include("Phone number can't be blank", "Phone number is invalid")
+        end
+
+        it 'phone_numberが11桁でなければ購入できない' do
+          @order_form.phone_number = " 123456789123"
+          @order_form.valid?
+          expect(@order_form.errors.full_messages).to include("Phone number is invalid")
+        end
```
```
phone_numberが英数字混合では購入できない
```
上記のテストも追加しましょう。

tatsuki-kato reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、確認をお願いします！

fujiyasu0522 committed:
修正1

テストコード
https://gyazo.com/d6f9e54407924bf19a15620c5b489e00

tatsuki-kato commented on app/models/order_form.rb:
```diff
@@ -0,0 +1,19 @@
+    validates :block_name
+    validates :postal_code,      format: { with: /\A\d{3}[-]\d{4}\z/ }
+    validates :delivery_area_id, numericality: { other_than: 0 }
+    validates :phone_number,     format: { with: /\A\d{11}\z/ }
```
```rb
validates :phone_number,     format: { with: /\A\d{11}\z/ }
```
上記ですと11桁の数字のみしか電話番号が保存できない仕様になってしまいます。

Trelloの実装条件にあるように
```
- 電話番号にはハイフンは不要で、11桁以内であること（09012345678となる）
```
を実装しましょう。

tatsuki-kato commented on spec/factories/order_forms.rb:
```diff
@@ -0,0 +1,13 @@
+    block_name        {'1-1-1'}
+    building          {'建物ビル'}
+    phone_number      {'09022223333'} 
+    user_id           { 1 }
+    item_id           { 1 }
```
```rb
user_id           { 1 }
item_id           { 1 }
```
外部キーを直接FactoryBotに記載するのはいい実装とは言えません。
今回購入において商品やユーザーを用意する場合、
フォームオブジェクトのモデルテストファイルのbefore内でuserやitemのFactoryBotから生成しましょう。
理由としては以下になります。
```
①フォームオブジェクトはアソシエーションが使えないため、バリデーション にitem_id及びuser_idを記載している。

②実際のitem_id及びuser_idにはそれぞれの商品やユーザーの情報が入ることになる。

③
item_id { 1 }
user_id { 1 }
のように記述してしまうと、
常に「idの情報が1のみでidに紐づく商品やユーザーの情報がない」の状態でテストが実行されることになり、
実際の挙動と異なる。
```

ご不明な点がありましたら、コネクトにてメンターにご相談ください！

tatsuki-kato commented on spec/models/order_form_spec.rb:
```diff
@@ -0,0 +1,101 @@
+require 'rails_helper'
+
+RSpec.describe OrderForm, type: :model do
+  before do
+    @order_form = FactoryBot.build(:order_form)
+  end
```
こちらの中にuserとitemのインスタンスをFactoryBotから保存した形で生成し、@order_formの中にユーザーと商品それぞれの情報が入るように実装しましょう。

■ヒント
```rb
user = #FactroyBotのcreateメソッドを使ってユーザーのインスタンスを保存した形で生成する。
item = #FactroyBotのcreateメソッドを使って商品のインスタンスを保存した形で生成する。
@order_form = FactoryBot.build(:order_form, user_id: [定義したuserが持つid], item_id: [定義したitemが持つid])
```

tatsuki-kato reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、確認をお願いします！

fujiyasu0522 committed:
修正2

https://gyazo.com/10f0e75734b44f3eae7cfcb9c13a6ebf

https://gyazo.com/1dc24a6ca055e8871e6c497aa12443ef

d-kazu4k commented on app/views/items/show.html.erb:
```diff
@@ -30,7 +30,7 @@
         <%= link_to "削除", item_path(@item.id), method: :delete, class:"item-destroy" %>
       <% else %>
     <%# 商品が売れていない場合はこちらを表示しましょう %>
```
```
<%# 商品が売れていない場合はこちらを表示しましょう %>
```
実装案内用のコメントアウトは削除しましょう！
もし必要であれば、第三者でもわかるように「残す目的」「なぜ必要か」という理由などを、コメントとして書いておくことをおすすめいたします。
他にも不要なコメントアウトがあれば、削除しましょう！

d-kazu4k commented on spec/factories/order_forms.rb:
```diff
@@ -0,0 +1,13 @@
+    block_name        {'1-1-1'}
+    building          {'建物ビル'}
+    phone_number      {"09012345678"} 
+    association :user
+    association :item
```
```ruby
    association :user
    association :item
```
上記は削除しましょう！
※フォームオブジェクトはテーブルを持たないモデルのため、アソシエーションが組めないからです。

d-kazu4k replied:
https://github.com/fujiyasu0522/furima-33357/pull/9/files#r648197788
上記ですが、下記ファイル内の記述も削除しましょう！
```
spec/factories/order_forms.rb
```

d-kazu4k reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いいたします。

fujiyasu0522 committed:
修正3

d-kazu4k commented on app/models/order_form.rb:
```diff
@@ -0,0 +1,19 @@
+  include ActiveModel::Model
+  attr_accessor :token, :postal_code, :delivery_area_id, :city_name, :block_name, :building, :phone_number, :item_id, :user_id
+
+  with_options presence: true do
```
```ruby
with_options presence: true do
```
`token`に対するバリデーションも追加しましょう！

d-kazu4k reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いいたします。

fujiyasu0522 committed:
修正4

d-kazu4k reviewed (commented):
レビュー依頼ありがとうございます！
LGTMとさせていただきます。実装お疲れさまでした！

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き頑張っていきましょう！
